components/os/state/state.go

```diff
@@ -0,0 +1,163 @@
+	start := time.Now()
+	rows, err := db.QueryContext(ctx, fmt.Sprintf(`SELECT boot_id, boot_unix_seconds FROM %s ORDER BY boot_unix_seconds ASC`, DeprecatedTableNameMachineBootIDs))
+	sqlite.RecordSelect(time.Since(start).Seconds())
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, nil
+		}
+		// e.g., "no such table: machine_boot_ids"
+		if strings.Contains(err.Error(), "no such table") {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer rows.Close()
+
+	events := []BootEvent{}
+	for rows.Next() {
+		ev := BootEvent{}
+		if err := rows.Scan(&ev.BootID, &ev.Timestamp); err != nil {
+			return nil, err
+		}
+		events = append(events, ev)
+	}
+	if len(events) == 0 {
+		return nil, nil
+	}
+	return events, nil
+}
+
+func CreateTable(ctx context.Context, db *sql.DB) error {
+	existing, err := readFromDeprecatedTable(ctx, db)
+	if err != nil {
+		log.Logger.Warnw("failed to read from deprecated table", "error", err)
+	}
+	if _, err := db.ExecContext(ctx, fmt.Sprintf(`DROP TABLE IF EXISTS %s`, DeprecatedTableNameMachineBootIDs)); err != nil {
+		return err
+	}
+
+	_, err = db.ExecContext(ctx, fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s (
+	%s TEXT PRIMARY KEY,
+	%s INTEGER
+);`, TableNameBootEvents, ColumnBootID, ColumnTimestamp))
+	if err != nil {
+		return err
+	}
+
+	for _, event := range existing {
+		log.Logger.Warnw("found previous boot event in deprecated table", "boot_id", event.BootID, "timestamp", event.Timestamp)
+		if err := InsertBootID(ctx, db, event.BootID, time.Unix(event.Timestamp, 0)); err != nil {
+			log.Logger.Warnw("failed to insert boot id", "error", err)
+		}
+	}
+
+	return nil
+}
+
+// Returns an empty string and no error if no boot ID is found.
+func GetLastBootID(ctx context.Context, db *sql.DB) (string, error) {
+	start := time.Now()
+	row := db.QueryRowContext(ctx, fmt.Sprintf("SELECT %s FROM %s ORDER BY %s DESC LIMIT 1", ColumnBootID, TableNameBootEvents, ColumnTimestamp))
+	sqlite.RecordSelect(time.Since(start).Seconds())
+
+	var bootID string
+	err := row.Scan(&bootID)
+	if err == sql.ErrNoRows {
+		return "", nil
+	}
+	return bootID, err
+}
+
+func GetFirstBootID(ctx context.Context, db *sql.DB) (string, error) {
+	start := time.Now()
+	row := db.QueryRowContext(ctx, fmt.Sprintf("SELECT %s FROM %s ORDER BY %s ASC LIMIT 1", ColumnBootID, TableNameBootEvents, ColumnTimestamp))
+	sqlite.RecordSelect(time.Since(start).Seconds())
+
+	var bootID string
+	err := row.Scan(&bootID)
+	if err == sql.ErrNoRows {
+		return "", nil
+	}
+	return bootID, err
+}
+
+func InsertBootID(ctx context.Context, db *sql.DB, bootID string, ts time.Time) error {
+	start := time.Now()
+	_, err := db.ExecContext(ctx, fmt.Sprintf("INSERT INTO %s (%s, %s) VALUES (?, ?)",
+		TableNameBootEvents,
+		ColumnBootID,
+		ColumnTimestamp),
+		bootID, ts.UTC().Unix())
+	sqlite.RecordInsertUpdate(time.Since(start).Seconds())
+
+	return err
+}
+
+type BootEvent struct {
```

Comment:
should this be in event.go? also we SHOULD NOT persist state into sqlite. state is meant to be transient.

Reply:
Yeah wasn't a good file name. Renaming.